api/server.go

```diff
@@ -74,22 +91,62 @@ func (srv *Server) Serve() error {
 	if err != nil && !strings.HasSuffix(err.Error(), "use of closed network connection") {
 		return err
 	}
+	return nil
+}
 
-	// safely close each module
-	if srv.cs != nil {
-		srv.cs.Close()
+// Close closes the Server's listener, causing the HTTP server to shut down.
+func (srv *Server) Close() error {
+	var errStrs []string
+
+	// Close the listener, which will cause Server.Serve() to return.
+	if err := srv.listener.Close(); err != nil {
+		errStrs = append(errStrs, fmt.Sprintf("listener err: %v", err))
 	}
-	if srv.gateway != nil {
-		srv.gateway.Close()
+
+	// Wait for Server.Serve() to exit. We wait so that it's guaranteed that the
+	// server has completely closed after Close() returns. This is particularly
+	// useful during testing so that we don't exit a test before Serve() finishes.
+	srv.wg.Wait()
+
+	// Safely close each module.
+	if srv.host != nil {
+		if err := srv.host.Close(); err != nil {
```

Comment:
try this:
```go
modules := []io.Closer{srv.host, srv.explorer, srv.miner...}
for _, m : = range modules {
    if m != nil {
        if err := m.Close(); err != nil {
            errs = append(errs, err)
        }
    }
}
```
although, then you lose the name in the error message. I suppose you could use a map for that, though...